### PR TITLE
Agent-level Metrics upgrade

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Work in progress. Upgrading agents to use the new Metrics object.

**Testing steps**
CI.

### CODE REVIEW

File: parlai/core/metrics.py [hunk 1]
```diff
@@ -84,7 +88,10 @@ def __repr__(self) -> str:
         return f'{self.__class__.__name__}({self.value():.4g})'
 
     def __float__(self) -> float:
-        return self.value()
+        return float(self.value())
+
+    def __int__(self) -> int:
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Pretty helpful, actually

File: parlai/core/metrics.py [hunk 2]
```diff
@@ -239,6 +295,23 @@ def compute(guess: str, answers: List[str], k: int = 4) -> Optional['BleuMetric'
         return BleuMetric(score)
 
 
+class FairseqBleuMetric(AverageMetric):
+    @staticmethod
+    def compute_many(
+        guess: torch.Tensor, answers: torch.Tensor, pad_idx, end_idx, unk_idx
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: nit: types on the idx args?

File: parlai/scripts/eval_model.py [hunk 3]
```diff
@@ -52,15 +52,6 @@ def setup_args(parser=None):
     parser.add_argument('-ne', '--num-examples', type=int, default=-1)
     parser.add_argument('-d', '--display-examples', type='bool', default=False)
     parser.add_argument('-ltim', '--log-every-n-secs', type=float, default=2)
-    parser.add_argument(
```
Comment (EricMichaelSmith) [inline parlai/scripts/eval_model.py]: Hmm, it's a bit unclear from `aggregate_named_reports()` alone, but I assume that it's doing micro-averaging since that's the point of tracking both the numerator and denominator? In any case, removing this flag indicates to me that it's clearly not being used much, ha

Comment (stephenroller) [reply parlai/scripts/eval_model.py]: Yeah we're explicitly doing micro averaging always now (correctly). @emilydinan are you okay with that?

File: parlai/utils/distributed.py [hunk 4]
```diff
@@ -79,6 +79,18 @@ def is_primary_worker():
     return not is_distributed() or dist.get_rank() == 0
 
 
+def get_rank():
```
Comment (stephenroller) [inline parlai/utils/distributed.py]: @mwillwork  this is what i was referring to

File: parlai/core/metrics.py [hunk 5]
```diff
@@ -300,92 +373,30 @@ def normalize_answer(s):
...
-        return reports[0]
-    # multiple tasks, aggregate metrics
-    metrics = {}
-    exs = {}
-    total_report = {'tasks': {}}
-    # collect metrics from all reports
-    for i, report in enumerate(reports):
-        total_report['tasks'][tasks[i]] = report
-        for metric, val in report.items():
-            if metric == 'exs':
```
Comment (stephenroller) [inline parlai/core/metrics.py]: Might be causing me problems in multitasking?

File: parlai/core/metrics.py [hunk 6]
```diff
@@ -300,92 +373,30 @@ def normalize_answer(s):
...
-        m['tasks'][task_id] = task_report
-        total += task_report.get('exs', 0)
-        found_any = False
-        for k in sums.keys():
-            if k in task_report:
-                sums[k] += task_report[k]
-                found_any = True
-        if found_any:
-            num_tasks += 1
-    m['exs'] = total
```
Comment (stephenroller) [inline parlai/core/metrics.py]: here too

File: .circleci/config.yml [hunk 7]
```diff
@@ -150,7 +150,7 @@ jobs:
       - <<: *installtorchcpu
       - run:
           name: Data tests
-          command: python -m coverage run -m pytest --junitxml=test-results/junit.xml -m data
+          command: python -m coverage run -m pytest --junitxml=test-results/junit.xml -m data -v
```
Comment (stephenroller) [inline .circleci/config.yml]: I'll undo these changes, they need to go elsewhere.

Comment (EricMichaelSmith) [inline .circleci/config.yml]: Hmm - what's the -v flag for here?

File: parlai/agents/seq2seq/seq2seq.py [hunk 8]
```diff
@@ -206,9 +206,9 @@ def build_model(self, states=None):
     def build_criterion(self):
         # set up criteria
         if self.opt.get('numsoftmax', 1) > 1:
-            return nn.NLLLoss(ignore_index=self.NULL_IDX, reduction='sum')
+            return nn.NLLLoss(ignore_index=self.NULL_IDX, reduction='none')
```
Comment (EricMichaelSmith) [inline parlai/agents/seq2seq/seq2seq.py]: Oy - I assume the `Metric.many()` operation is handling the reduction now

File: parlai/core/logs.py [hunk 9]
```diff
@@ -68,6 +69,8 @@ def log_metrics(self, setting, step, report):
         for k, v in report.items():
             if isinstance(v, numbers.Number):
                 self.writer.add_scalar(f'{setting}/{k}', v, global_step=step)
+            elif isinstance(v, Metric):
```
Comment (EricMichaelSmith) [inline parlai/core/logs.py]: It *feels* like we shouldn't need to have a separate condition for this because it's mostly a repeat of above, but I don't know how to combine the two conditions without obfuscating what Metric does

File: parlai/core/metrics.py [hunk 10]
```diff
@@ -136,7 +167,8 @@ def __add__(self, other: Optional['SumMetric']) -> 'SumMetric':
         if other is None:
             return self
         full_sum = self._sum + other._sum
-        return SumMetric(sum_=full_sum)
+        # always keep the same return type
+        return type(self)(sum_=full_sum)
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Great catch here

File: parlai/core/metrics.py [hunk 11]
```diff
@@ -300,92 +377,30 @@ def normalize_answer(s):
...
-
-
-def aggregate_metrics(reporters):
+    # reporters is a list of teachers or worlds
+    m: Dict[str, Metric] = {}
+    for task_id, task_report in named_reports.items():
+        for each_metric, value in task_report.items():
+            m[each_metric] = m.get(each_metric, None) + value
+            if len(named_reports) > 1:
+                m[f'{task_id}/{each_metric}'] = value
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: So the choice of a slash notation in the key string here confuses me a bit. Why not have a second dictionary within the first?

File: parlai/core/metrics.py [hunk 12]
```diff
@@ -300,92 +377,30 @@ def normalize_answer(s):
...
-                sums[k] += task_report[k]
-                found_any = True
-        if found_any:
-            num_tasks += 1
-    m['exs'] = total
-    m['accuracy'] = 0
-    if num_tasks > 0:
-        for k in sums.keys():
-            m[k] = round_sigfigs(sums[k] / num_tasks, 4)
+            m[each_metric] = m.get(each_metric) + value
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Great to see the conciseness here!

File: parlai/core/metrics.py [hunk 13]
```diff
@@ -444,7 +456,7 @@ def _drain_queue(self):
             return
         while not self._queue.empty():
             try:
-                yield self._queue.get()
+                yield self._queue.get(timeout=0.05)
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Not familiar with this queue. Is this timeout to avoid race conditions or something like that?

File: parlai/core/teachers.py [hunk 14]
```diff
@@ -262,8 +265,7 @@ def __init__(self, opt, shared=None):
 
         # set up batching
         self.bsz = opt.get('batchsize', 1)
-        self.batchindex = opt.get('batchindex', 0)
-        self.use_batch_act = False  # Batch act disabled by default
+        self.use_batch_act = False
```
Comment (EricMichaelSmith) [inline parlai/core/teachers.py]: Okay, don't immediately understand why we can get away with deleting .batch_act() altogether here. And we still have a flag for .use_batch_act even though the method has been deleted?

File: parlai/core/teachers.py [hunk 15]
```diff
@@ -1944,6 +1904,10 @@ def shutdown(self):
         for t in self.tasks:
             t.shutdown()
 
+    def update_counters(self):
```
Comment (EricMichaelSmith) [inline parlai/core/teachers.py]: Lots of new code here to update counters, and it seems to ultimately be syncing metrics. I suppose this is needed because the whole idea of this working is predicated on local metrics always being up-to-date?

File: parlai/core/torch_generator_agent.py [hunk 16]
```diff
@@ -623,15 +554,24 @@ def compute_loss(self, batch, return_output=False):
...
-        target_tokens = notnull.long().sum().item()
-        correct = ((batch.label_vec == preds) * notnull).sum().item()
-        self.metrics['correct_tokens'] += correct
-        self.metrics['nll_loss'] += loss.item()
-        self.metrics['num_tokens'] += target_tokens
-        loss /= target_tokens  # average loss per token
+        target_tokens = notnull.long().sum(dim=-1)
+        correct = ((batch.label_vec == preds) * notnull).sum(dim=-1)
+
+        if not skip_metrics:
```
Comment (EricMichaelSmith) [inline parlai/core/torch_generator_agent.py]: Can you explain when we'd need to skip metrics here vs. not?

File: parlai/core/torch_ranker_agent.py [hunk 17]
```diff
@@ -457,13 +449,16 @@ def eval_step(self, batch):
...
+            self.record_local_metric('loss', AverageMetric.many(loss))
+            ranks_m = []
+            mrrs_m = []
             for b in range(batchsize):
                 rank = (ranks[b] == label_inds[b]).nonzero()
                 rank = rank.item() if len(rank) == 1 else scores.size(1)
-                self.metrics['rank'] += 1 + rank
-                self.metrics['mrr'] += 1.0 / (1 + rank)
+                ranks_m.append(1 + rank)
+                mrrs_m.append(1.0 / (1 + rank))
```
Comment (EricMichaelSmith) [inline parlai/core/torch_ranker_agent.py]: I'd be curious to understand this calculation, but that's beyond the scope of this PR =P

File: parlai/core/torch_ranker_agent.py [hunk 18]
```diff
@@ -755,41 +750,6 @@ def share(self):
         shared['optimizer'] = self.optimizer
         return shared
 
-    def reset_metrics(self):
```
Comment (EricMichaelSmith) [inline parlai/core/torch_ranker_agent.py]: This is not needed anymore because of the syncing operations, perhaps?

File: parlai/scripts/train_model.py [hunk 19]
```diff
@@ -445,6 +431,9 @@ def save_model(self, suffix=None):
             except KeyboardInterrupt:
                 pass
 
+    def _safe_report(self, report):
```
Comment (EricMichaelSmith) [inline parlai/scripts/train_model.py]: This might be more generally useful as a helper function beyond train_model.py, I wonder

File: parlai/utils/misc.py [hunk 20]
```diff
@@ -282,6 +283,10 @@ def log(self, done, total, report=None):
             log dict contains pairs of all items to log, which includes
             percentage complete and projected time left if total > 0
         """
+        from parlai.core.metrics import Metric  # delay import to prevent circular dep
```
Comment (EricMichaelSmith) [inline parlai/utils/misc.py]: Ugh - necessary, I'm sure, but circular deps scare me

File: parlai/utils/misc.py [hunk 21]
```diff
@@ -282,6 +283,10 @@ def log(self, done, total, report=None):
...
+
+        if isinstance(done, Metric):
+            done = done.value()
```
Comment (EricMichaelSmith) [inline parlai/utils/misc.py]: Again - wonder if this pattern can be abstracted away

File: parlai/utils/typing.py [hunk 22]
```diff
@@ -31,3 +31,5 @@ class _Shared(Dict[str, Any]):
 Ints and floats are clearly scalars, and torch.Tensors can be represented by a scalar if
 Tensor.numel() == 1. Used as input type for classes derived from Metric.
 """
+
+TVector = Union[List[TScalar], torch.Tensor]
```
Comment (EricMichaelSmith) [inline parlai/utils/typing.py]: Wonder if we'd need to add numpy arrays here too?

File: projects/image_chat/transresnet_multimodal/transresnet_multimodal.py [hunk 23]
```diff
@@ -423,13 +423,17 @@ def report(self):
         """
         Report per-dialogue round metrics.
         """
-        m = {k: {} for k in ["first_round", "second_round", "third_round+"]}
+        m = {}
```
Comment (EricMichaelSmith) [inline projects/image_chat/transresnet_multimodal/transresnet_multimodal.py]: oh god - getting into the weeds of the individual projects now. I suppose people will know if their project hasn't been converted because it'll stop working?

File: tests/test_eval_model.py [hunk 24]
```diff
@@ -116,19 +116,15 @@ def test_multitasking_metrics(self):
...
+        task2_acc = valid['integration_tests:multiturnCandidate/accuracy']
         total_acc = valid['accuracy']
         # task 2 is 4 times the size of task 1
-        self.assertAlmostEqual(
-            total_acc,
-            (task1_acc + 4 * task2_acc) / 5,
-            4,
-            'Task accuracy is averaged incorrectly',
+        self.assertEqual(
+            total_acc, task1_acc + task2_acc, 'Task accuracy is averaged incorrectly',
```
Comment (EricMichaelSmith) [inline tests/test_eval_model.py]: The division-by-5 test doesn't work? Sometimes complicated tests are good

File: tests/test_train_model.py [hunk 25]
```diff
@@ -37,15 +37,11 @@ def test_multitasking_metrics(self):
...
+        task1_acc = valid['integration_tests:candidate/accuracy']
+        task2_acc = valid['integration_tests:multiturnCandidate/accuracy']
         total_acc = valid['accuracy']
-        # task 2 is 4 times the size of task 1
-        self.assertAlmostEqual(
-            total_acc,
-            (task1_acc + 4 * task2_acc) / 5,
-            4,
-            'Task accuracy is averaged incorrectly',
+        self.assertEqual(
```
Comment (EricMichaelSmith) [inline tests/test_train_model.py]: Pretty repetitive code compared to test_seq2seq.py above

File: parlai/core/metrics.py [hunk 26]
```diff
@@ -394,54 +409,90 @@ class Metrics(object):
...
+    def flush(self):
+        """
+        Clear the local buffer and push it on.
+        """
+        if self._threadsafe and self._buffer:
+            try:
+                self._queue.put(self._buffer[:])
+                self._buffer.clear()
+            except queue.Full:
+                raise SystemError(
```
Comment (stephenroller) [inline parlai/core/metrics.py]: this can go.

File: parlai/core/metrics.py [hunk 27]
```diff
@@ -98,6 +109,14 @@ def __lt__(self, other: Any) -> bool:
         else:
             return self.value() < other
 
+    def __sub__(self, other: Any) -> float:
+        """
+        Used heavily for assertAlmostEqual.
+        """
+        if not isinstance(other, float):
+            raise TypeError('Metrics.__sub__ is intentionally limited to floats.')
```
Comment (klshuster) [inline parlai/core/metrics.py]: why? is this purely because of the `assertAlmostEqual`?

File: parlai/core/metrics.py [hunk 28]
```diff
@@ -160,12 +192,40 @@ def __add__(self, other: Optional['AverageMetric']) -> 'AverageMetric':
...
+        self._value = self.as_number(value)
+
+    def value(self) -> float:
+        return self._value
+
+    def __add__(self, other: Optional['FixedMetric']) -> 'FixedMetric':
+        if other is None:
+            return self
+        if self != other:
+            raise ValueError(f'Trying to two unequal fixed metrics: {self} != {other}')
```
Comment (klshuster) [inline parlai/core/metrics.py]: nit: `Trying to add two...`

File: parlai/core/teachers.py (hunk 14 above)
Comment (klshuster) [reply parlai/core/teachers.py]: can we just get rid of `use_batch_act` altogether now?

File: parlai/core/torch_agent.py [hunk 29]
```diff
@@ -882,28 +890,59 @@ def build_lr_scheduler(self, states=None, hard_reset=False):
                 self.scheduler.get_initial_number_training_updates()
             )
 
+    def _control_local_metrics(self, enabled: bool = False, disabled: bool = False):
+        """
+        Used to temporarily disable local metrics.
+
+        This is useful for things like when you need to call super(), but
+        prevent the parent from recording some some metric. For example,
```
Comment (klshuster) [inline parlai/core/torch_agent.py]: nit: "some some"

File: parlai/core/torch_classifier_agent.py [hunk 30]
```diff
@@ -247,13 +248,11 @@ def train_step(self, batch):
         labels = self._get_labels(batch)
         scores = self.score(batch)
         loss = self.criterion(scores, labels)
+        self.record_local_metric('loss', AverageMetric.many(loss))
```
Comment (klshuster) [inline parlai/core/torch_classifier_agent.py]: does this not keep the loss tensor around?

File: parlai/core/torch_classifier_agent.py [hunk 31]
```diff
@@ -286,8 +285,8 @@ def eval_step(self, batch):
         else:
             labels = self._get_labels(batch)
             loss = self.criterion(scores, labels)
-            self.metrics['loss'] += loss.item()
-            self.metrics['examples'] += len(batch.text_vec)
+            self.record_local_metric('loss', AverageMetric.many(loss))
```
Comment (klshuster) [inline parlai/core/torch_classifier_agent.py]: related to above comment, i thought part of the reason we were calling `.item()` was to get the scalar value rather than keep the tensor around

Comment (stephenroller) [reply parlai/core/torch_classifier_agent.py]: Metrics automatically calls item

File: parlai/core/metrics.py (hunk 27 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: Yup!
